src/csharp/Grpc.Core/ChannelCredentials.cs

```diff
@@ -104,40 +105,74 @@ internal override ChannelCredentialsSafeHandle CreateNativeCredentials()
         }
     }
 
+    /// <summary>
+    /// Callback invoked with the expected targetHost and the peer's certificate.
+    /// If a non-zero value is returned by this callback then it is treated as a
+    /// verification failure. Invocation of the callback is blocking, so any
+    /// implementation should be light-weight.
+    /// </summary>
+    /// <param name="context">The <see cref="T:Grpc.Core.VerifyPeerContext"/> associated with the callback</param>
+    /// <returns>true if verification succeeded, false otherwise.</returns>
+    /// Note: experimental API that can change or be removed without any prior notice.
+    public delegate bool VerifyPeerCallback(VerifyPeerContext context);
+
     /// <summary>
     /// Client-side SSL credentials.
     /// </summary>
     public sealed class SslCredentials : ChannelCredentials
     {
         readonly string rootCertificates;
         readonly KeyCertificatePair keyCertificatePair;
+        readonly VerifyPeerCallbackInternal verifyPeerCallback;
+        readonly VerifyPeerDestruct verifyPeerDestruct;
+        readonly GCHandle gcHandle;
```

Comment:
you don't need two GCHandle. 
- either you can use just one native callback (with more params)
- or you can alloc GCHandle on the SslCredentials class instance itself (the instance references the delegates so they cannot be GC'd).

Reply:
Switched to a single native callback.